# Computational Intelligence Homework 2

# Used libraries

In [1]:
import pandas as pd
import numpy as np
from tabulate import tabulate
import 

# Fetching the data

In [2]:
trainset = pd.read_csv('../Dataset/train.csv')
testset = pd.read_csv('../Dataset/test.csv')

## Demonstrating the data
the train set

In [3]:
trainset

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

the test set

In [4]:
testset

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

## Showing how much data is missing

In [5]:
trainset.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [6]:
trainset = trainset.fillna(method='pad')
# data imputation, the numerical missing data is replaced by the mean of the dataset and the categorical data is 
# replaced by the mod of the dataset.
trainset.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          1
Embarked       0
dtype: int64

In [7]:
testset.isna().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

In [8]:
testset = testset.fillna(method='pad')
# data imputation, the numerical missing data is replaced by the mean of the dataset and the categorical data is 
# replaced by the mod of the dataset.
testset.isna().sum()

PassengerId     0
Pclass          0
Name            0
Sex             0
Age             0
SibSp           0
Parch           0
Ticket          0
Fare            0
Cabin          12
Embarked        0
dtype: int64

# Feature Engineeing, preprocessing our data
removing unnecessary features and keeping the important ones. 

***NOTE: the cabin number had 147 categories and 687 missing data so I decided to remove this feature.***

In [9]:
# removing Name, Ticket and Cabin features.
trainset.drop(['Name', 'Ticket', 'Cabin'], axis=1, inplace=True)
trainset

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare  \
0              1         0       3    male  22.0      1      0   7.2500   
1              2         1       1  female  38.0      1      0  71.2833   
2              3         1       3  female  26.0      0      0   7.9250   
3              4         1       1  female  35.0      1      0  53.1000   
4              5         0       3    male  35.0      0      0   8.0500   
..           ...       ...     ...     ...   ...    ...    ...      ...   
886          887         0       2    male  27.0      0      0  13.0000   
887          888         1       1  female  19.0      0      0  30.0000   
888          889         0       3  female  19.0      1      2  23.4500   
889          890         1       1    male  26.0      0      0  30.0000   
890          891         0       3    male  32.0      0      0   7.7500   

    Embarked  
0          S  
1          C  
2          S  
3          S  
4          S  
..       ...  
886        S  
887        S  
888        S  
889        C  
890        Q  

[891 rows x 9 columns]

In [10]:
# removing Name, Ticket and Cabin features.
testset.drop(['Name', 'Ticket', 'Cabin'], axis=1, inplace=True)
testset

PassengerId  Pclass     Sex   Age  SibSp  Parch      Fare Embarked
0            892       3    male  34.5      0      0    7.8292        Q
1            893       3  female  47.0      1      0    7.0000        S
2            894       2    male  62.0      0      0    9.6875        Q
3            895       3    male  27.0      0      0    8.6625        S
4            896       3  female  22.0      1      1   12.2875        S
..           ...     ...     ...   ...    ...    ...       ...      ...
413         1305       3    male  28.0      0      0    8.0500        S
414         1306       1  female  39.0      0      0  108.9000        C
415         1307       3    male  38.5      0      0    7.2500        S
416         1308       3    male  38.5      0      0    8.0500        S
417         1309       3    male  38.5      1      1   22.3583        C

[418 rows x 8 columns]

### changing categorical data to numerical using one hot encoding method.

in below firstly, we counted the number of categories for each column.

In [11]:
print('Sex categories =',trainset['Sex'].unique())
print('Embarked categories =',trainset['Embarked'].unique())

Sex categories = ['male' 'female']
Embarked categories = ['S' 'C' 'Q']


In [12]:
# replacing category columns with boolean columns
sex_column = trainset['Sex'].to_numpy()
male_column = (sex_column == 'male') * 1 # this multiplication converts boolean to integer values
female_column = (sex_column == 'female') * 1 # this multiplication converts boolean to integer values
trainset.insert(3, 'Male', male_column)
trainset.insert(3, 'Female', female_column)
trainset[['Female', 'Male', 'Sex']]

Female  Male     Sex
0         0     1    male
1         1     0  female
2         1     0  female
3         1     0  female
4         0     1    male
..      ...   ...     ...
886       0     1    male
887       1     0  female
888       1     0  female
889       0     1    male
890       0     1    male

[891 rows x 3 columns]

In [13]:
trainset.drop('Sex', axis=1, inplace=True)
trainset

PassengerId  Survived  Pclass  Female  Male   Age  SibSp  Parch     Fare  \
0              1         0       3       0     1  22.0      1      0   7.2500   
1              2         1       1       1     0  38.0      1      0  71.2833   
2              3         1       3       1     0  26.0      0      0   7.9250   
3              4         1       1       1     0  35.0      1      0  53.1000   
4              5         0       3       0     1  35.0      0      0   8.0500   
..           ...       ...     ...     ...   ...   ...    ...    ...      ...   
886          887         0       2       0     1  27.0      0      0  13.0000   
887          888         1       1       1     0  19.0      0      0  30.0000   
888          889         0       3       1     0  19.0      1      2  23.4500   
889          890         1       1       0     1  26.0      0      0  30.0000   
890          891         0       3       0     1  32.0      0      0   7.7500   

    Embarked  
0          S  
1          C  
2          S  
3          S  
4          S  
..       ...  
886        S  
887        S  
888        S  
889        C  
890        Q  

[891 rows x 10 columns]

### doing the same thing for Embarked column

In [14]:
embarked_column = trainset['Embarked'].to_numpy()
s_column = (embarked_column == 'S') * 1 # this multiplication converts boolean to integer values
c_column = (embarked_column == 'C') * 1 # this multiplication converts boolean to integer values
q_column = (embarked_column == 'Q') * 1 # this multiplication converts boolean to integer values
trainset.insert(9, 'Embarked_S', s_column)
trainset.insert(9, 'Embarked_C', c_column)
trainset.insert(9, 'Embarked_Q', q_column)
trainset.drop('Embarked', axis=1, inplace=True)

### doing the exact same thing for the trainset

In [15]:
# replacing category columns with boolean columns
sex_column = testset['Sex'].to_numpy()
male_column = (sex_column == 'male') * 1 # this multiplication converts boolean to integer values
female_column = (sex_column == 'female') * 1 # this multiplication converts boolean to integer values
testset.insert(3, 'Male', male_column)
testset.insert(3, 'Female', female_column)
testset.drop('Sex', axis=1, inplace=True)

embarked_column = testset['Embarked'].to_numpy()
s_column = (embarked_column == 'S') * 1 # this multiplication converts boolean to integer values
c_column = (embarked_column == 'C') * 1 # this multiplication converts boolean to integer values
q_column = (embarked_column == 'Q') * 1 # this multiplication converts boolean to integer values
testset.insert(8, 'Embarked_S', s_column)
testset.insert(8, 'Embarked_C', c_column)
testset.insert(8, 'Embarked_Q', q_column)
testset.drop('Embarked', axis=1, inplace=True)

PassengerId  Pclass  Female  Male   Age  SibSp  Parch      Fare  \
0            892       3       0     1  34.5      0      0    7.8292   
1            893       3       1     0  47.0      1      0    7.0000   
2            894       2       0     1  62.0      0      0    9.6875   
3            895       3       0     1  27.0      0      0    8.6625   
4            896       3       1     0  22.0      1      1   12.2875   
..           ...     ...     ...   ...   ...    ...    ...       ...   
413         1305       3       0     1  28.0      0      0    8.0500   
414         1306       1       1     0  39.0      0      0  108.9000   
415         1307       3       0     1  38.5      0      0    7.2500   
416         1308       3       0     1  38.5      0      0    8.0500   
417         1309       3       0     1  38.5      1      1   22.3583   

     Embarked_Q  Embarked_C  Embarked_S  
0             1           0           0  
1             0           0           1  
2             1           0           0  
3             0           0           1  
4             0           0           1  
..          ...         ...         ...  
413           0           0           1  
414           0           1           0  
415           0           0           1  
416           0           0           1  
417           0           1           0  

[418 rows x 11 columns]

# Extracting the output column from trainset

In [16]:
survived_vec = trainset['Survived'].to_numpy()
trainset.drop('Survived', axis=1, inplace=True)

# Cross-Validation 
this is for deciding the hyper parameter K in KNN algorithm.
for this matter we divide the trainset into 4 folds and validate on one fold and train on the others in each step and take average on errors to announce the validation on the parameter. we take this method when we lack training data and the test set is separated before hand.

***NOTE: we used sklearn libraries for this section***

the distance between the data is defined by their cosine similarity.

In [18]:
def neighbours_of(input):
    pass

this gives the output of the sample requested from the dataset.

In [19]:
def supervise(n):
    pass

because this is a classification problem we should use majority function or the mod of the K nearest data to determine what is the predicted class of the input.

In [20]:
def majority(neibs):
    count = [0,0]
    for n in neibs:
        count[supervise(n)] +=1
    if(count[0] > count[1]):
        return 0;
    else:
        return 1;

In [21]:
def KNN(input, k):
    neibs = neighbours_of(input, k)
    prediction = majority(neibs)
    return prediction